pkg/accounting/accounting.go

```diff
@@ -0,0 +1,210 @@
+	balance  int64  // amount that the peer owes us if positive, our debt if negative
+	reserved uint64 // amount currently reserved for active peer interaction
+}
+
+// Options for accounting
+type Options struct {
+	DisconnectThreshold uint64
+	Logger              logging.Logger
+	Store               storage.StateStorer
+}
+
+// Accounting is the main implementation of the accounting interface
+type Accounting struct {
+	balancesMu sync.Mutex // mutex for accessing the balances map
+	balances   map[string]*PeerBalance
+	logger     logging.Logger
+	store      storage.StateStorer
+
+	disconnectThreshold uint64 // the debt threshold at which we will disconnect from a peer
+}
+
+// NewAccounting creates a new Accounting instance with the provided options
+func NewAccounting(o Options) *Accounting {
+	return &Accounting{
+		balances:            make(map[string]*PeerBalance),
+		disconnectThreshold: o.DisconnectThreshold,
+		logger:              o.Logger,
+		store:               o.Store,
+	}
+}
+
+// Reserve reserves a portion of the balance for peer
+func (a *Accounting) Reserve(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	// the previously reserved balance plus the new price is the maximum amount paid if all current operations are successful
+	// since we pay this we have to reduce this (positive quantity) from the balance
+	// the disconnectThreshold is stored as a positive value which is why it must be negated prior to comparison
+	if balance.balance-int64(balance.reserved+price) < -int64(a.disconnectThreshold) {
+		return fmt.Errorf("cannot afford operation with peer %v", peer.String())
+	}
+
+	balance.reserved += price
+
+	return nil
+}
+
+// Release releases reserved funds
+func (a *Accounting) Release(peer swarm.Address, price uint64) {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		a.logger.Error("Releasing balance for peer that is not loaded")
+		return
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	if price > balance.reserved {
+		// If Reserve and Release calls are always paired this should never happen
+		a.logger.Error("Releasing more balance than was reserved for peer")
+		balance.reserved = 0
+	} else {
+		balance.reserved -= price
+	}
+}
+
+// Debit increases the balance we have with the peer
+func (a *Accounting) Debit(peer swarm.Address, price uint64) error {
+	balance, err := a.getPeerBalance(peer)
+	if err != nil {
+		return err
+	}
+
+	balance.lock.Lock()
+	defer balance.lock.Unlock()
+
+	nextBalance := balance.balance + int64(price)
+
+	a.logger.Tracef("debiting peer %v with price %d, new balance is %d", peer, price, nextBalance)
+
+	if nextBalance >= int64(a.disconnectThreshold) {
+		// peer to much in debt
+		return p2p.NewDisconnectError(fmt.Errorf("disconnect threshold exceeded for peer %s", peer.String()))
+	}
+
+	err = a.store.Put(a.balanceKey(peer), nextBalance)
```

Comment:
It does not matter if its the same Kademlia, assuming your overlay address does not change you will probably have some of the same peers and if you did not persist you balances might be out of sync from the beginning (this is especially bad if you were already close to payment threshold before restarting). What's not so important is a few seconds of lost accounting since the difference in payment and disconnect threshold should allow for that.